library/x509write_crt.c

```diff
@@ -293,6 +293,43 @@ int mbedtls_x509write_crt_set_key_usage( mbedtls_x509write_cert *ctx,
     return( 0 );
 }
 
+int mbedtls_x509write_crt_set_ext_key_usage( mbedtls_x509write_cert *ctx,
+                                             const mbedtls_asn1_sequence *exts )
+{
+    unsigned char buf[256];
+    unsigned char *c = buf + sizeof(buf);
+    int ret;
+    size_t len = 0;
+    const mbedtls_asn1_sequence *last_ext = NULL;
+    const mbedtls_asn1_sequence *ext;
+
+    memset( buf, 0, sizeof(buf) );
+
+    /* We need at least one extension: SEQUENCE SIZE (1..MAX) OF KeyPurposeId */
+    if( exts == NULL )
+        return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
+
+    /* Iterate over exts backwards, so we write them out in the requested order */
```

Comment:
Ah, now I understand this comment - it's commenting on `cert_write.c`, which does indeed add the extensions in the reverse order specified on the command line.

Now, `programs/` are meant to be example programs, not the main interface to the library, so it seems the "proper" way of doing this would be for this function to add the extensions in the order specified within `exts`, and for `cert_write.c` to not reverse the list it creates (yes, I am aware that this happens because of the simple way it builds a linked list).

How much do we care about doing this correctly?

Reply:
I think the library behaviour is correct here. The library does add the extensions in the order of the linked list that is passed to it (note that the `mbedtls_asn1_write_...` calls write to the end of the buffer and decrement the pointer).

cert_write does construct the list backwards, so the extensions end up backwards. It's easy enough to fix so I'll update.